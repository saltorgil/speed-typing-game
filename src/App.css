@import url(./fonts/fontfaces.css);

html, body{
    margin : 0;
    padding : 0;
    text-align: center;
    background: black;
    color: #00b800;
    font-family: 'astron_boy';
    font-size: 100%
}

.main-container{
    font-size: 2rem;
}