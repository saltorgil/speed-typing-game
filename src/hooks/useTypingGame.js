import  {useState, useEffect, useRef} from "react";

function useTypingGame(start_game_time){

    const [text, setText] = useState('')
    const [timeRemaining, setTimeRemaining] = useState(start_game_time)
    const [isRunning, setIsRunning] = useState(false)
    const [numWords, setNumWords] = useState(0)

    const [allQuotes, setAllQuotes] = useState([])
    const [randomQuote, setRandomQuote] = useState('')


    const refTextarea = useRef(null);

    const handleTextarea = (event) =>{
      setText(event.target.value)
    }
  
    const countTextWords = (text) => 
      text
        .trim()
        .split(' ')
        .filter(word=>word!=='')
        .length
  
    const startGame = () => {
      setIsRunning(true)
      setTimeRemaining(start_game_time)
      setText('')
      refTextarea.current.disabled = false
      refTextarea.current.focus()
    }
  
    const endGame = () =>{
      setIsRunning(false);
      setNumWords(countTextWords(text));
      getRandomQuote();
    }

    const getRandomQuote = () => {
      const randNum = Math.floor(Math.random() * allQuotes.length)
      setRandomQuote(allQuotes[randNum])
    }
  
    useEffect(()=>{
        document.title = "Speed Typing Game";
  
      if (isRunning && timeRemaining > 0 ){
        setTimeout(()=>{
          setTimeRemaining(prevTime => prevTime -1) 
        },1000)
      }
      else if (timeRemaining === 0) {
        endGame();
        console.log("juego terminado")
      }
    // eslint-disable-next-line
    },[timeRemaining, isRunning])


    useEffect(()=>{

      async function randomQuote() {
        const response = await fetch('https://api.quotable.io/quotes?' + new URLSearchParams({
            maxLength: 55,
            skip: Math.floor(Math.random() * 50)
        }))

        const data = await response.json()

        const {results} = data
        const arrayQuotes = results.map(result => result.content + " - " + result.author)
        setAllQuotes(arrayQuotes);
      }
      
      randomQuote()
    },[])

    // eslint-disable-next-line
    useEffect(()=>getRandomQuote(), [allQuotes])


    return {text, timeRemaining, isRunning, randomQuote, numWords, refTextarea, handleTextarea, countTextWords, startGame}
}

export default useTypingGame