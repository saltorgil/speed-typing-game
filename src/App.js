import React from 'react';
import useTypingGame from './hooks/useTypingGame';
import './css/app.css';
import logo from './images/logox2.png';

import audio from './intergalactic.ogg';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faVolumeUp, faVolumeMute } from '@fortawesome/free-solid-svg-icons';

function App() {
  const {
    text,
    timeRemaining,
    isRunning,
    randomQuote,
    numWords,
    refTextarea,
    refAudio,
    handleTextarea,
    countTextWords,
    startGame,
    volumeOn,
    handleVolumeOn,
  } = useTypingGame(5);

  return (
    <div className='main'>
      <audio ref={refAudio} loop autoPlay hidden>
        <source src={audio} type='audio/ogg' />
      </audio>
      <div className='main-container'>
        {/* Main container header */}
        <div className='main-container-header'>
          <div className='main-container-header-logo'>
            <img src={logo} alt='logo'></img>
          </div>
          <FontAwesomeIcon
            onClick={handleVolumeOn}
            className='main-container-header-volume'
            icon={volumeOn ? faVolumeUp : faVolumeMute}
            size='lg'
          />
        </div>

        {/* Main container separation */}
        <hr className='main-container-hr' />

        {/* Main container body */}
        <div className='main-container-body'>
          <h1>How fast do you type in only 5 seconds?</h1>
          <h3>
            TRY THIS <span>&#9654; </span>
            {randomQuote}
          </h3>
          <textarea
            className='main-container-body-textarea'
            onChange={handleTextarea}
            value={text}
            disabled={!isRunning}
            ref={refTextarea}
          />
          <h3>Time remaining: {timeRemaining}</h3>
          <button
            className='main-container-body-button sheen'
            onClick={startGame}
            disabled={isRunning}>
            Start
          </button>
          <h1 id='main-container-click' onClick={countTextWords}>
            Word Count: {numWords}
          </h1>
        </div>
      </div>
    </div>
  );
}

export default App;
