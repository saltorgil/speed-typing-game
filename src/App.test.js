import React from 'react';
import { shallow, mount } from 'enzyme';
import App from './App';
import useTypingGame from './hooks/useTypingGame';
import { countTextWords } from './hooks/useTypingGame';

// Test component UI components
describe('<App /> with no props', () => {
  const container = shallow(<App />);
  it('should match the snapshot', () => {
    expect(container.html()).toMatchSnapshot();
  });

  it('should have an texarea', () => {
    expect(container.find('textarea').length).toEqual(1);
  });

  it('should have proper props for textarea', () => {
    expect(container.find('textarea').props()).toEqual({
      className: 'main-container-body-textarea',
      onChange: expect.any(Function),
      disabled: true,
      value: '',
    });
  });

  it('should have a button', () => {
    expect(container.find('button').length).toEqual(1);
  });

  it('should have proper props for button field', () => {
    expect(container.find('button').props()).toEqual({
      className: 'main-container-body-button sheen',
      onClick: expect.any(Function),
      disabled: false,
      children: 'Start',
    });
  });

  it('should have a main-container-click', () => {
    expect(container.find('#main-container-click').length).toEqual(1);
  });

  it('should have proper props for button field', () => {
    expect(container.find('#main-container-click').props()).toEqual({
      id: 'main-container-click',
      onClick: expect.any(Function),
      children: expect.anything(),
    });
  });
});

// Test write textarea
describe('<App /> textarea writing after click counts', () => {
  it('should set the textarea value on change event', () => {
    const container = shallow(<App />);

    container.find('textarea').simulate('change', {
      target: {
        value: 'Lorem Ipsum',
      },
    });
    expect(container.find('textarea').prop('value')).toEqual('Lorem Ipsum');
  });

  it('should set Word count to 2', () => {
    const container = mount(<App />);

    container.find('button').simulate('click');

    container.find('textarea').simulate('change', {
      target: {
        value: 'Lorem Ipsum',
      },
    });

    setTimeout(() => {
      expect(container.find('#main-container-click').text()).toEqual(
        'Word Count: 2'
      );
    }, 5000);
  });
});

// Unit test util function
describe('countTextWords tests', function () {
  test('empty string should count 0', () => {
    expect(countTextWords('')).toEqual(0);
  });

  test('"Lorem ipsum" sentence should count 2', () => {
    expect(countTextWords('Lorem ipsun')).toEqual(2);
  });
  test('" Lorem  ipsum " sentence with two or more spaces should count 2', () => {
    expect(countTextWords(' Lorem  ipsun ')).toEqual(2);
  });
});
