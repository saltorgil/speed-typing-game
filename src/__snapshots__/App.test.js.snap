// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> with no props should match the snapshot 1`] = `"<div class=\\"main\\"><audio loop=\\"\\" autoplay=\\"\\" hidden=\\"\\"><source src=\\"intergalactic.ogg\\" type=\\"audio/ogg\\"/></audio><div class=\\"main-container\\"><div class=\\"main-container-header\\"><div class=\\"main-container-header-logo\\"><img src=\\"logox2.png\\" alt=\\"logo\\"/></div><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"volume-mute\\" class=\\"svg-inline--fa fa-volume-mute fa-w-16 fa-lg main-container-header-volume\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M215.03 71.05L126.06 160H24c-13.26 0-24 10.74-24 24v144c0 13.25 10.74 24 24 24h102.06l88.97 88.95c15.03 15.03 40.97 4.47 40.97-16.97V88.02c0-21.46-25.96-31.98-40.97-16.97zM461.64 256l45.64-45.64c6.3-6.3 6.3-16.52 0-22.82l-22.82-22.82c-6.3-6.3-16.52-6.3-22.82 0L416 210.36l-45.64-45.64c-6.3-6.3-16.52-6.3-22.82 0l-22.82 22.82c-6.3 6.3-6.3 16.52 0 22.82L370.36 256l-45.63 45.63c-6.3 6.3-6.3 16.52 0 22.82l22.82 22.82c6.3 6.3 16.52 6.3 22.82 0L416 301.64l45.64 45.64c6.3 6.3 16.52 6.3 22.82 0l22.82-22.82c6.3-6.3 6.3-16.52 0-22.82L461.64 256z\\"></path></svg></div><hr class=\\"main-container-hr\\"/><div class=\\"main-container-body\\"><h1>How fast do you type in only 5 seconds?</h1><h3>TRY THIS <span>â–¶ </span></h3><textarea class=\\"main-container-body-textarea\\" disabled=\\"\\"></textarea><h3>Time remaining: 5</h3><button class=\\"main-container-body-button sheen\\">Start</button><h1 id=\\"main-container-click\\">Word Count: 0</h1></div></div></div>"`;
